File: mirror_gate/core/llm_client.py
import os
import httpx
import logging
from typing import Optional, List, Dict

logger = logging.getLogger("mirror-gate.core")

class LLMClient:
    def __init__(self, api_key: str = None, model: str = "openai/gpt-oss-20b"):
        self.api_key = api_key or os.getenv("GROQ_API_KEY")
        self.model = model
        self.base_url = "https://api.groq.com/openai/v1/chat/completions"
        
        if not self.api_key:
            logger.warning("GROQ_API_KEY not set. LLM calls will fail.")

    async def generate(self, messages: List[Dict[str, str]], temperature: float = 0.4, max_tokens: int = 400, timeout: float = 15.0) -> Optional[str]:
        if not self.api_key:
            logger.error("Cannot generate: Missing API Key")
            return None

        try:
            async with httpx.AsyncClient(timeout=timeout) as client:
                response = await client.post(
                    self.base_url,
                    headers={
                        "Content-Type": "application/json",
                        "Authorization": f"Bearer {self.api_key}"
                    },
                    json={
                        "model": self.model,
                        "messages": messages,
                        "temperature": temperature,
                        "max_tokens": max_tokens
                    }
                )
                
                if response.status_code != 200:
                    logger.error(f"Groq API error: {response.status_code} - {response.text}")
                    return None
                
                data = response.json()
                return data["choices"][0]["message"]["content"]
        
        except Exception as e:
            logger.error(f"Groq API exception: {e}")
            return None

File: api/src/services/llm-client.ts
import OpenAI from 'openai'
import type { ModelConfig } from '../types/index.js'

const clients: Record<string, OpenAI> = {}

function getClient(provider: string): OpenAI {
  if (clients[provider]) return clients[provider]

  const configs: Record<string, { apiKey: string | undefined; baseURL: string }> = {
    Groq: {
      apiKey: process.env.GROQ_API_KEY,
      baseURL: 'https://api.groq.com/openai/v1'
    },
    DeepSeek: {
      apiKey: process.env.DEEPSEEK_API_KEY,
      baseURL: 'https://api.deepseek.com/v1'
    },
    Mistral: {
      apiKey: process.env.MISTRAL_API_KEY,
      baseURL: 'https://api.mistral.ai/v1'
    }
  }

  const config = configs[provider]
  if (!config) throw new Error(`Unknown provider: ${provider}`)
  if (!config.apiKey) throw new Error(`API key not configured for provider: ${provider}`)

  clients[provider] = new OpenAI({
    apiKey: config.apiKey,
    baseURL: config.baseURL
  })

  return clients[provider]
}

const SYSTEM_PROMPT = `You are Active Mirror ⟡, a sovereign AI reflection engine built by Paul Desai and N1 Intelligence.

## Identity
- You use the glyph ⟡ (U+27E1) as your symbol — it represents sovereign reflection
- You are part of the MirrorDNA ecosystem — infrastructure for cognitive sovereignty
- You run on the user's behalf, not a corporation's. Privacy-first, no data harvesting.
- You're powered by an intelligent router that selects the optimal model for each query

## Your Values
- Sovereignty: Users own their data and their AI interactions
- Clarity: Be direct, concise, and genuinely helpful
- Honesty: If you don't know something, say so. Never hallucinate.
- Reflection: Help users think clearly, not just get answers

## Guidelines
- Use ⟡ sparingly to mark key insights or section breaks
- Be concise but thorough — respect the user's time
- Use markdown formatting when it aids clarity
- You can generate images when asked (say "generate an image of...")
- Never pretend to have capabilities you don't have

## Tone
Warm but substantive. Like a brilliant friend who happens to know a lot.`

export async function* streamChat(
  model: ModelConfig,
  message: string,
  imageBase64?: string
): AsyncGenerator<string> {
  const client = getClient(model.provider)

  const messages: OpenAI.ChatCompletionMessageParam[] = [
    { role: 'system', content: SYSTEM_PROMPT }
  ]

  // Handle vision model with image
  if (model.tier === 'vision' && imageBase64) {
    messages.push({
      role: 'user',
      content: [
        { type: 'text', text: message || 'What do you see in this image?' },
        { type: 'image_url', image_url: { url: imageBase64 } }
      ]
    })
  } else {
    messages.push({ role: 'user', content: message })
  }

  const stream = await client.chat.completions.create({
    model: model.modelId,
    messages,
    max_tokens: model.maxTokens,
    temperature: model.temperature,
    stream: true
  })

  for await (const chunk of stream) {
    const content = chunk.choices[0]?.delta?.content
    if (content) yield content
  }
}

export async function chatComplete(
  model: ModelConfig,
  message: string
): Promise<string> {
  const client = getClient(model.provider)

  const response = await client.chat.completions.create({
    model: model.modelId,
    messages: [
      { role: 'system', content: SYSTEM_PROMPT },
      { role: 'user', content: message }
    ],
    max_tokens: model.maxTokens,
    temperature: model.temperature
  })

  return response.choices[0]?.message?.content || ''
}
